x/authz/keeper/keeper.go

```diff
@@ -0,0 +1,191 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	proto "github.com/gogo/protobuf/proto"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/authz/types"
+)
+
+type Keeper struct {
+	storeKey sdk.StoreKey
+	cdc      codec.BinaryMarshaler
+	router   *baseapp.MsgServiceRouter
+}
+
+// NewKeeper constructs a message authorization Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc codec.BinaryMarshaler, router *baseapp.MsgServiceRouter) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+		router:   router,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// getAuthorizationGrant returns grant between granter and grantee for the given msg type
+func (k Keeper) getAuthorizationGrant(ctx sdk.Context, actor []byte) (grant types.AuthorizationGrant, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(actor)
+	if bz == nil {
+		return grant, false
+	}
+	k.cdc.MustUnmarshalBinaryBare(bz, &grant)
+	return grant, true
+}
+
+func (k Keeper) update(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, updated types.Authorization) error {
+	actor := types.GetActorAuthorizationKey(grantee, granter, updated.MethodName())
+	grant, found := k.getAuthorizationGrant(ctx, actor)
+	if !found {
+		return fmt.Errorf("no authorization found")
+	}
+
+	msg, ok := updated.(proto.Message)
+	if !ok {
+		return fmt.Errorf("cannot proto marshal %T", updated)
+	}
+
+	any, err := codectypes.NewAnyWithValue(msg)
+	if err != nil {
+		return err
+	}
+
+	grant.Authorization = any
+	store := ctx.KVStore(k.storeKey)
+	store.Set(actor, k.cdc.MustMarshalBinaryBare(&grant))
+	return nil
+}
+
+// DispatchActions attempts to execute the provided messages via authorization
+// grants from the message signer to the grantee.
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, serviceMsgs []sdk.ServiceMsg) (*sdk.Result, error) {
+	var msgResult *sdk.Result
+	var err error
+	for _, serviceMsg := range serviceMsgs {
+		signers := serviceMsg.GetSigners()
+		if len(signers) != 1 {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "authorization can be given to msg with only one signer")
+		}
+		granter := signers[0]
+		if !bytes.Equal(granter, grantee) {
+			authorization, _ := k.GetAuthorization(ctx, grantee, granter, serviceMsg.MethodName)
```

Comment:
I would add here: `if expriationTime.Before(ctx.Block().Time) { revoke }`

Reply:
This is handled inside `GetAuthorization`, that's the reason to invoke `revoke` from `GetAuthorization`

Reply:
Yeah, see https://github.com/cosmos/cosmos-sdk/pull/7629#discussion_r561917781. Mentally, I don't think we should modify the store in a getter function. I think putting it here is okay.

The other solution would be to rename that function to e.g. `GetOrRevokeAuthorization`

Reply:
Yes, but querying the authrozation would still return an authorization even if it is expired (if we revoke in `DispatchActions`). May be `GetOrRevokeAuthorization` would be better choice.